lib/improver/spotdata_new/apply_lapse_rate.py

```diff
@@ -0,0 +1,168 @@
+                        all cubes.
+        """
+        def _get_attributes(cube):
+            """Build dictionary of attributes that match the
+            self.grid_metadata_identifier."""
+
+            attributes = cube.attributes
+            attributes = {k: v for (k, v) in attributes.items()
+                          if self.grid_metadata_identifier in k}
+            return attributes
+
+        def _compare_attributes(attributes, reference_attributes):
+            """Compare keys and values of attributes with reference_attributes.
+            Raise an exception if they do not match exactly."""
+            try:
+                match = ((attributes.items() & reference_attributes.items()) ==
+                         reference_attributes.items())
+            except TypeError:
+                match = ((attributes.viewitems() &
+                          reference_attributes.viewitems()) ==
+                         reference_attributes.viewitems())
+            if match is not True:
+                raise ValueError('Cubes do not share the metadata identified '
+                                 'by the grid_metadata_identifier ({})'.format(
+                                     self.grid_metadata_identifier))
+
+        # Allow user to bypass cube comparison by setting identifier to None.
+        if self.grid_metadata_identifier is None:
+            return
+
+        reference_attributes = _get_attributes(cubes[0])
+        for cube in cubes:
+            attributes = _get_attributes(cube)
+            _compare_attributes(attributes, reference_attributes)
+
+    def process(self, spot_data_cube, neighbour_cube, gridded_lapse_rate_cube):
+        """
+        Extract lapse rates from the appropriate grid points and apply them to
+        the spot extracted temperatures.
+
+        Args:
+            spot_data_cube (iris.cube.Cube):
+                The spot data cube of temperatures extracted from the gridded
+                fields. These temperatures will have been extracted using the
+                same neighbour_cube and neighbour_selection_method that are
+                being used here.
+            neighbour_cube (iris.cube.Cube):
+                The neighbour_cube that contains the grid coordinates at which
+                lapse rates should be extracted and the vertical displacement
+                between those grid points on the model orography and the spot
+                data sites actual altitudes.
+            gridded_lapse_rate_cube (iris.cube.Cube):
+                A cube of temperature lapse rates on the same grid as that from
+                which the spot data temperatures were extracted.
+        Returns:
+            new_spot_cube (iris.cube.Cube):
+                A copy of the input spot_data_cube with the data modified by
+                the lapse rates to give a better representation of the site's
+                temperatures.
+        """
```

Comment:
It might be nice here if the actual equation that we are applying was added to the docstring, as it's a little bit hidden by the other more technical operations, so just something like e.g.:
```
lapse_rate_adjusted_temperatures = temperatures + lapse_rate * vertical_displacement
```

Reply:
Added.